ports/esp32/machine_i2c.c

```diff
@@ -49,92 +49,159 @@
+    esp_err_t err = i2c_master_probe(self->bus_handle, addr, 1000);
+    if (err != ESP_OK) {
+        return -MP_ENODEV;          /* No device at address, return immediately */
+    }
+    /* 1. Create a temporary device handle for this transaction */
+    i2c_device_config_t dev_cfg = {
+        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+        .device_address = addr,
+        .scl_speed_hz = 100000,   /* Use bus frequency */
+    };
+    i2c_master_dev_handle_t dev_handle;
+    err = i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &dev_handle);
+    if (err != ESP_OK) {
+        return -MP_ENODEV;
+    }
+
     int data_len = 0;
 
+    /* 2. If WRITE1 segment exists, perform the write first */
     if (flags & MP_MACHINE_I2C_FLAG_WRITE1) {
-        i2c_master_start(cmd);
-        i2c_master_write_byte(cmd, addr << 1, true);
-        i2c_master_write(cmd, bufs->buf, bufs->len, true);
+        if (bufs->len) {
+            err = i2c_master_transmit(dev_handle, bufs->buf, bufs->len, 1000);                 /* Block with 1 s timeout */
+            if (err != ESP_OK) {
+                goto cleanup;
+            }
+        }
         data_len += bufs->len;
         --n;
         ++bufs;
     }
+    if (flags & MP_MACHINE_I2C_FLAG_READ) {
+        /* 3. Main loop: remaining segments */
+        for (; n--; ++bufs) {
+            if (bufs->len == 0) {
+                continue;
+            }
+            err = i2c_master_receive(dev_handle, bufs->buf, bufs->len, 1000);
+            if (err != ESP_OK) {
+                break;
+            }
+
+            data_len += bufs->len;
+        }
+    } else {
+        // Write operation logic
+        size_t total_len = 0;
+        mp_machine_i2c_buf_t *original_bufs = bufs;  // Save original pointer
+        size_t yuann = n;
 
-    i2c_master_start(cmd);
-    i2c_master_write_byte(cmd, addr << 1 | (flags & MP_MACHINE_I2C_FLAG_READ), true);
+        // Calculate total length
+        for (; n--; ++bufs) {
+            total_len += bufs->len;
+        }
 
-    for (; n--; ++bufs) {
-        if (flags & MP_MACHINE_I2C_FLAG_READ) {
-            i2c_master_read(cmd, bufs->buf, bufs->len, n == 0 ? I2C_MASTER_LAST_NACK : I2C_MASTER_ACK);
-        } else {
-            if (bufs->len != 0) {
-                i2c_master_write(cmd, bufs->buf, bufs->len, true);
-            }
+        // Reset pointer
+        bufs = original_bufs;
+        // Reset n
+        n = yuann;
+        // Dynamically allocate write_buf
+        uint8_t *write_buf = (uint8_t *)malloc(total_len);
+        if (write_buf == NULL) {
+            return -MP_ENOMEM;
+        }
+
+        // Copy data into write_buf
+        size_t index = 0;
+        for (; n--; ++bufs) {
+            memcpy(write_buf + index, bufs->buf, bufs->len);
+            index += bufs->len;
+        }
+
+        // Transmit data
+        err = i2c_master_transmit(dev_handle, write_buf, total_len, 1000);
+        if (err != ESP_OK) {
```

Comment:
This test can be removed, since the label cleanup: follows anyhow. And with the test write_buf would not be free'd in case of an error, causing memory leakage.